File: decision-making/events/event_market.py
import json
import os
import time
import asyncio
from pathlib import Path
from events.event import Event
from events.event_handler import EventHandler
from events.event_callbacks import EventCallbacks

class EventMarket:
    def __init__(self):
        self.event_list = []
        self.file_path = Path(__file__).parent.parent / 'logs' / 'event_market.json'
        self.file_path.parent.mkdir(parents=True, exist_ok=True)
        self._initialize_file()
        EventCallbacks.set_event_market(self)

    def _initialize_file(self):
        if not self.file_path.exists() or self.file_path.stat().st_size == 0:
            with open(self.file_path, 'w') as f:
                json.dump([], f)

    def add_event_listener(self, name, callback):
        self.event_list.append(EventHandler(name, callback))

    async def emit_event(self, name, data):
        new_event = Event(name, data)
        try:
            events = await self._read_events_from_file()
            events.append(new_event.__dict__)
            await self._write_events_to_file(events)
        except Exception as e:
            print(f"Error emitting event: {e}")

    async def _read_events_from_file(self):
        with open(self.file_path, 'r') as f:
            try:
                return json.load(f)
            except json.JSONDecodeError:
                return []

    async def _write_events_to_file(self, events):
        with open(self.file_path, 'w') as f:
            json.dump(events, f, indent=2)

    async def _remove_event_by_id(self, event_id):
        events = await self._read_events_from_file()
        events = [event for event in events if event['id'] != event_id]
        await self._write_events_to_file(events)

    async def check_for_updates(self):
        events = await self._read_events_from_file()
        for event in events:
            for handler in self.event_list:
                if handler.name == event['name']:
                    await handler.callback(event['data'])
                    await self._remove_event_by_id(event['id'])

    async def start(self, interval=1):
        while True:
            await self.check_for_updates()
            await asyncio.sleep(interval)

File: data-collection/src/events/event_market.ts
import Event from './event';
import EventHandler from './event_handler';
import { promises as fs } from 'fs';
import path from 'path';

class EventMarket {
    eventList: EventHandler[] = [];
    filePath: string;

    constructor() {
        this.filePath = path.join(__dirname, '../../logs', 'event_market.json');

        (async () => {
            await this.initializeFile();

            while (true) {
                await this.checkForUpdates();
                await new Promise(resolve => setTimeout(resolve, 1000));
            }
        })();
    }

    async initializeFile(): Promise<void> {
        try {
            await fs.access(this.filePath);
        } catch (error : any) {
            if (error.code === 'ENOENT') {
                await this.writeEventsToFile([]);
            } else {
                console.error('Error checking file existence:', error);
            }
        } finally {
            await this.resetEventMarketFile();
        }
    }

    async resetEventMarketFile(): Promise<void> {
        try {
            await this.writeEventsToFile([]);
        } catch (error) {
            console.error('Error resetting event market file:', error);
        }
    }

    addEventListener(name: string, callback: Function): void {
        this.eventList.push(new EventHandler(name, callback));
    }

    async executeEvent(name: string, data: any): Promise<void> {
        const eventsToExecute = this.eventList.filter(event => event.name == name);    
        for (const eventHandler of eventsToExecute) {
            eventHandler.callback(data);
        }
    }

    async checkForUpdates(): Promise<void> {
        try {
            const events = await this.readEventsFromFile();
            for (const event of events) {
                if (this.eventList.some(e => e.name === event.name)) {
                    await this.executeEvent(event.name, event.data);
                    await this.removeEventById(event.id);
                }
            }
        } catch (error) {
            this.handleFileReadError(error);
        }
    }

    async readEventsFromFile(): Promise<Event[]> {
        const data = await fs.readFile(this.filePath, 'utf8');
        return JSON.parse(data || '[]');
    }

    async writeEventsToFile(events: Event[]): Promise<void> {
        await fs.writeFile(this.filePath, JSON.stringify(events, null, 2), 'utf8');
    }

    async removeEventById(id: string): Promise<void> {
        try {
            const events = await this.readEventsFromFile();
            const updatedEvents = events.filter(event => event.id !== id);
            await this.writeEventsToFile(updatedEvents);
        } catch (error) {
            console.error('Error removing event by ID:', error);
        }
    }

    async emitEvent(name: string, data: any): Promise<void> {
        const newEvent = new Event(name, data);

        try {
            const events = await this.readEventsFromFile();
            events.push(newEvent);
            await this.writeEventsToFile(events);
        } catch (error : any) {
            if (error.code === 'ENOENT') {
                await this.writeEventsToFile([newEvent]);
            } else {
                console.error('Error emitting event:', error);
            }
        }
    }

    handleFileReadError(error: any): void {
        if (error.code === 'ENOENT') {
            console.log('Event file not found, skipping check.');
        } else {
            console.error('Error reading event file:', error);
        }
    }
}

export default EventMarket;
